client/src/client.rs

```diff
@@ -231,17 +263,56 @@ pub trait RpcApi: Sized {
         T::query(&self, &id)
     }
 
+    // this method could be moved outside of the trait
+    fn get_network_info_request(&self) -> Request<json::GetNetworkInfoResult> {
+        Request {
+            method: "getnetworkinfo",
+            args: vec![],
+            converter: &|raw| converter_json::<json::GetNetworkInfoResult>(raw),
+        }
+    }
+    #[cfg(not(feature = "async"))]
     fn get_network_info(&self) -> Result<json::GetNetworkInfoResult> {
-        self.call("getnetworkinfo", &[])
+        self.get_network_info_request().sync(self)
+    }
+    #[cfg(feature = "async")]
+    async fn get_network_info(&self) -> Result<json::GetNetworkInfoResult> {
+        self.get_network_info_request().async(self)
+    }
+    // this is some way through which it would be possible to have an API with both sync and async
+    // methods when async is supported
+    #[cfg(feature = "both-sync-async")]
+    fn get_network_info_sync(&self) -> Result<json::GetNetworkInfoResult> {
+        self.get_network_info_request().sync(self)
+    }
+    fn batch_get_network_info<R>(&self, batch: &mut Batch<R>) where Self: Sized {
+        self.get_network_info_request().batch(batch);
     }
 
-    fn version(&self) -> Result<usize> {
-        #[derive(Deserialize)]
-        struct Response {
-            pub version: usize,
+    fn version_request(&self) -> Request<usize> {
+        Request {
+            method: "getnetworkinfo",
+            args: vec![],
+            converter: &|raw| {
+                #[derive(Deserialize)]
+                struct Response {
+                    pub version: usize,
+                }
+                let ret: Response = serde_json::from_str(raw.get())?;
+                Ok(ret.version)
+            },
         }
-        let res: Response = self.call("getnetworkinfo", &[])?;
-        Ok(res.version)
+    }
+    #[cfg(not(feature = "async"))]
```

Comment:
Ideally the API should support both, no? That way two dependencies setting different feature flags wouldn't cause compile failure.

Reply:
spoke with @stevenroose about this, we think there is a very clean approach for this by impl future for the request type directly. he's prototyping it :)